src/coins.cpp

```diff
@@ -301,16 +306,20 @@ bool CCoinsViewCache::Flush() {
 
 bool CCoinsViewCache::Sync()
 {
-    bool fOk = base->BatchWrite(cacheCoins, hashBlock, /*erase=*/false);
+    const auto fOk{base->BatchWrite(m_flagged_head.second.Next(), hashBlock, /*erase=*/false)};
     // Instead of clearing `cacheCoins` as we would in Flush(), just clear the
     // FRESH/DIRTY flags of any coin that isn't spent.
-    for (auto it = cacheCoins.begin(); it != cacheCoins.end(); ) {
+    for (auto it{m_flagged_head.second.Next()}; it != nullptr; ) {
         if (it->second.coin.IsSpent()) {
+            // We need to get next entry before we erase this entry.
+            // Since cacheCoins owns the entry, it will be deallocated before we
+            // can access the next entry if we erase first.
+            const auto next_entry{it->second.Next(/*clear_flags=*/true)};
```

Comment:
nit, you could move that line in front of the `if`, then remove the else, like so:

```cpp
const auto next_entry{it->second.Next(/*clear_flags=*/true)};
if (it->second.coin.IsSpent()) {
    cachedCoinsUsage -= it->second.coin.DynamicMemoryUsage();
    cacheCoins.erase(it->first);
}
it = next_entry;
```